src/engine/Account-Imap.cpp

```diff
@@ -0,0 +1,912 @@
+    {
+        int64_t existing_token_count = 0;
+
+        auto account_frame = qof_instance_get_slots (QOF_INSTANCE(imap->get_account()));
+
+        auto value = account_frame->get_slot(kvp_path.c_str());
+
+        // get the existing_token_count value
+        if (value != nullptr && value->get_type() == KvpValue::Type::INT64)
+            existing_token_count = value->get<int64_t>();
+
+        PINFO("found existing value of '%" G_GINT64_FORMAT "'", existing_token_count);
+
+        token_count = token_count + existing_token_count;
+    }
+    // Add or Update the entry based on guid
+    auto account_frame = qof_instance_get_slots (QOF_INSTANCE(imap->get_account()));
+
+    auto new_value = new KvpValue(token_count);
+
+    account_frame->set_path(kvp_path.c_str(), new_value);
+
+    /* Set a feature flag in the book for the change to use guid.
+     * This will prevent older GnuCash versions that don't support this feature
+     * from opening this file. */
+    gnc_features_set_used (imap->get_book(), GNC_FEATURE_GUID_BAYESIAN);
+}
+
+
+void
+GncImportMatchMap::add_account_bayes (GList* tokens, Account *acc)
+{
+    GList *current_token;
+
+    gchar *account_fullname;
+    gchar *guid_string;
+    std::string kvp_path;
+    std::string delim = "/";
+
+    ENTER(" ");
+
+    g_return_if_fail (acc != nullptr);
+    account_fullname = gnc_account_get_full_name (acc);
+    xaccAccountBeginEdit (m_account);
+
+    PINFO("account name: '%s'", account_fullname);
+    g_free (account_fullname);
+
+    guid_string = guid_to_string (xaccAccountGetGUID (acc));
+
+    /* process each token in the list */
+    for (current_token = g_list_first(tokens); current_token;
+            current_token = current_token->next)
+    {
+        /* Jump to next iteration if the pointer is not valid or if the
+                 string is empty. In HBCI import we almost always get an empty
+                 string, which doesn't work in the kvp loopkup later. So we
+                 skip this case here. */
+        if (!current_token->data || (*((char*)current_token->data) == '\0'))
+            continue;
+
+        std::string token_data = ((char*)current_token->data);
+
+        /* start off with one token for this account */
+        int64_t token_count = 1;
+
+        PINFO("adding token '%s'", token_data.c_str());
+
+        kvp_path = IMAP_FRAME_BAYES + delim + token_data + delim + (guid_string);
+
+        /* change the imap entry for the account */
+        change_imap_entry (this, kvp_path, token_count);
+    }
+    /* free up the account fullname and guid string */
+    qof_instance_set_dirty (QOF_INSTANCE (m_account));
+    xaccAccountCommitEdit (m_account);
+    g_free (guid_string);
+
+    LEAVE(" ");
+}
+
+/*****************************************************************************/
+
+GncImportMatchMap *
+gnc_account_imap_create_imap (Account *acc)
+{
+    if (!acc) return nullptr;
+
+    GncImportMatchMap *ImportMap = new GncImportMatchMap(acc);
```

Comment:
You can't use operator new for objects that you pass back to C code, C doesn't know how to free it.